### PR TITLE
chore: generate attestations during a release

### LINKED ISSUE #10848: Automatically update the `stable` tag after a release
**Describe the bug**

After a release on the current stable branch, we currently have to manually bump the tag. This results in folks not getting the latest manifests: https://github.com/argoproj/argo-cd/pull/10775#issuecomment-1271625320

**To Reproduce**

Do a release on the current stable minor release series and then note that the `stable` tag target does not change.

**Expected behavior**

The GitHub release action should be modified to bump the stable tag on a stable release.

That'll probably involve scripting that
1) checks what minor version `stable` currently points to
2) if the release is for the same minor version, update and push the tag at the end of the release

### PR DESCRIPTION
**Pull request by 34fathombelow**
## tldr
This PR generates attestations during a release using a new release process.  Each attestation contains a [SLSA Level 3 provenance](https://slsa.dev/spec/v0.1/requirements#provenance-requirements) which is then cryptographically signed. A SLSA [provenance](https://slsa.dev/provenance/v0.2) is generated for Argo CD container images and CLI binaries that are published during a release. These provenances are non-falsifable and generated using [slsa-github-generator](https://github.com/slsa-framework/slsa-github-generator). The release workflow needed to be refactored in order to provide a isolated environment for each step of the build process. This ensures a build **will not** influence another build, while also denying access to the required permissions to sign a artifact or generate a provenance.

## Why we need a provenance
Signatures only prove that a image was signed by someone with access to the container registry. A provenance provides metadata on who, what, where, when, and how a software artifact was generated. It is then cryptographically signed to provide an attestation. A provenance can be used in a CI/CD system or in a Kubernetes cluster to attest that the image was published by the intended source. 

## Reasoning behind the 2 step release process
In order to comply with higher [SLSA](https://slsa.dev/spec/v0.1/requirements#verified-history) requirements, commits to the release branch should be code reviewed and approved.  In particular during the release process the `manifest` and the `VERSION` file are updated.

## Notable changes to the release process

- Provenance for containers images
- Provenance for CLI binaries
- Container images and provenaces are signed using a keyless method that produce ephemeral certificates(we no longer manage private keys)
~~- CLI binaries are released in a compressed format (tar.gz/zip)~~
- GoReleaser used to produce CLI binaries, checksums, and auto generated release-notes
- `VERSION` file in master branch is updated automatically when a new release becomes GA
- The `stable tag` is automatically updated when the necessary requirements are met
- Pushing images to dockerhub during a release was removed.

## Notable files that involve the release process.

| File                               | Description                                                     |
|------------------------------------|-----------------------------------------------------------------|
|goreleaser.yaml                     |Config to build CLI binaries, checksums, release-notes           |
|.github/workflows/image-reuse.yaml  |Reusable workflow used to generate and sign container images     |
|.github/workflows/init-release.yaml |Used to generate the `manifest` and `VERSION` file  for a release                   |
|.github/workflows/release.yaml      |Build image, CLI binaries, provenances, sbom, post jobs          |

### image-reuse.yaml

- The reusable workflow can be used to publish or build images with multiple container registries(Quay,GHCR, dockerhub), and then sign them with cosign when an image is published.
- A GO version `must` be specified e.g. 1.19
- The image name for each registry *must* contain the tag. Note: multiple image names are allowed for each registry using a CSV type.
- Multiple platforms can be specified e.g. linux/amd64,linux/arm64
- Images are not published by default. A boolean value must be set to `true` to push images.
- An optional target can be specified.

| Inputs            | Description                         | Type        | Required | Defaults        |
|-------------------|-------------------------------------|-------------|----------|-----------------|
| go-version        | Version of Go to be used            | string      | true     | none            |
| quay_image_name   | Full image name and tag             | CSV, string | false    | none            |
| ghcr_image_name   | Full image name and tag             | CSV, string | false    | none            |
| docker_image_name | Full image name and tag             | CSV, string | false    | none            |
| platforms         | Platforms to build (linux/amd64)    | CSV, string | false    | linux/amd64     |
| push              | Whether to push a image to the registry | boolean     | false    | false           |
| target            | Target build stage                  | string      | false    | none            |

| Outputs     | Description                              | Type  |
|-------------|------------------------------------------|-------|
|image-digest | Image digest of image container created | string|

## Requirements for merging approver.

- Add label to repo `release`
- Add label to repo `test-multi-image`

## Testing
All new and modified files have been fully tested in a [public repo](https://github.com/34fathombelow/cosign_test).  Access to the repo can be requested for testing or trying out the new process.

## New release page images

image

image

image

 
Fixes https://github.com/argoproj/argo-cd/issues/10848

### CODE REVIEW

**Review by todaywasawesome — APPROVED** [review]
Excellent contribution! - I'm marking this approve though I would like to see the attestations pushed to Dockerhub aswell.

**Comment by 34fathombelow** [comment]
> Excellent contribution! - I'm marking this approve though I would like to see the attestations pushed to Dockerhub aswell.

@todaywasawesome I will implement and test this change.

**Comment by 34fathombelow** [comment]
Changes made:
1. Container images will not be pushed to quay and dockerhub along with the attestations
2. Release assets will not long be compressed and will retain our traditional naming schema

File: .goreleaser.yaml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    ## Release Signatures and Provenance
+
+    All Argo CD container images are signed by cosign.  A Provenance is generated for container images and CLI binaries which meet the SLSA Level 3 specifications. See the [documentation](https://argo-cd.readthedocs.io/en/stable/operator-manual/signed-release-assets) on how to verify.
+
+
+    ## Upgrading
+
+    If upgrading from a different minor version, be sure to read the [upgrading](https://argo-cd.readthedocs.io/en/stable/operator-manual/upgrading/overview/) documentation.
+  footer: |
+    **Full Changelog**: https://github.com/34fathombelow/cosign_test/compare/{{ .PreviousTag }}...{{ .Tag }}
```
Comment (34fathombelow) [inline .goreleaser.yaml:85]:
```suggestion
    **Full Changelog**: https://github.com/argoproj/cosign_test/argo-cd/{{ .PreviousTag }}...{{ .Tag }}
```

**Comment by crenshaw-dev** [comment]
I'm going to merge this right before we cut 2.7.0-rc1, on Mar. 27.

**Review by crenshaw-dev — APPROVED** [review]
Justin disabled workflows for older release branches, so we can actually merge now.

**Comment by jannfis** [comment]
Just a question to:

> VERSION file in master branch is updated automatically when a new release becomes GA

Should this not be updated once we branched the appropriate release branch, e.g. we branched off `release-2.7`, then VERSION should be bumped to _next_ major release, which would be `2.8.0`

**Comment by 34fathombelow** [comment]
> Just a question to:
> 
> > VERSION file in master branch is updated automatically when a new release becomes GA
> 
> Should this not be updated once we branched the appropriate release branch, e.g. we branched off `release-2.7`, then VERSION should be bumped to _next_ major release, which would be `2.8.0`

I agree, I will change this after the 2.7 release tomorrow.  I will enable the `VERSION` file to be updated when a RC1 release is performed.
